apps/api/views/metrics.py

```diff
@@ -0,0 +1,761 @@
+        bucket_quantity = 12
+
+        # Bucket intervals lie in between bucket values
+        bucket_bounds = []
+
+        for i in range(bucket_quantity + 1):
+            bucket_bound = round(minimum + (i * delta / bucket_quantity), 1)
+            bucket_bounds.append(bucket_bound)
+
+        i = 0
+        j = 0
+        bucket = 0
+        buckets = []
+        while j < len(bucket_bounds):
+            if i < len(sorted_sample) and sorted_sample[i].get('score') < bucket_bounds[j]:
+                if j != 0:
+                    bucket += 1
+                i += 1
+            else:
+                if j != 0:
+                    buckets.append(bucket)
+                    bucket = 0
+                j += 1
+
+        data = {
+            'values': buckets,
+            'labels': bucket_bounds,
+        }
+        return data
+
+
+class TimeSeriesObjectCreationQueryMixin:
+    time_series_model = None
+    time_series_creation_date_field_name = None
+
+    def time_series_query(self, count_field_name='count'):
+        output_fields = {
+            count_field_name: Count('pk'),
+            'date': F('datefield')
+        }
+        time_series_data = self.time_series_model.objects.dates(self.time_series_creation_date_field_name, 'year').values(**output_fields)
+        return time_series_data
+
+
+class ScorePerHWMixin:
+    score_per_hw_filter_name = None
+
+    def score_per_hw_query(self, filter_object, team_query=False):
+        sub_query_filter = {
+            self.score_per_hw_filter_name: filter_object,
+            'definition': OuterRef('pk'),
+        }
+        definition_filter = {
+            'klass': filter_object.klass
+        }
+        if team_query:
+            definition_filter['team_based'] = True
+        data = Definition.objects.filter(**definition_filter).annotate(
+            non_normalized_score=Max(Subquery(
+                Submission.objects.filter(**sub_query_filter).values('tracked_submissions__stored_score')[:1])
+            )
+        ).order_by('due_date').annotate(score=((F('non_normalized_score') - F('baseline_score')) / (F('target_score') - F('baseline_score')))).values('name', 'score')
+        return data
+
+
+class KlassScorePerHWMixin:
+    def klass_score_per_hw_query(self, **kwargs):
+        klass_pk = kwargs.get('klass_pk')
+        data = []
+        for definition in Definition.objects.filter(klass=klass_pk).values('pk', 'name', 'target_score', 'baseline_score').order_by('due_date'):
+            qs = StudentMembership.objects.filter(
+                klass=klass_pk
+            ).annotate(
+                max_score=Max(Subquery(
+                    Submission.objects.filter(
+                        id__in=OuterRef('submitted_homeworks'),
+                        definition=definition.get('pk')
+                    ).values('tracked_submissions__stored_score'))
+                )
+            ).values_list('max_score', flat=True)[:1]
+
+            sum = 0
```

Comment:
Overriding default `sum()` here. May want to change this variable name.

Reply:
Thank you. Done.